File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/session/ProcessObserver.java
/*
 * Copyright (c) 2015-2022 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.snowplow.internal.session;

import android.annotation.TargetApi;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ProcessLifecycleOwner;

import android.content.Context;
import android.os.Build;
import android.os.Handler;

import com.snowplowanalytics.snowplow.internal.utils.NotificationCenter;
import com.snowplowanalytics.snowplow.internal.tracker.Logger;

import java.util.HashMap;
import java.util.Map;

@RestrictTo(RestrictTo.Scope.LIBRARY)
@TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
public class ProcessObserver implements LifecycleObserver {
    private static final String TAG = ProcessObserver.class.getSimpleName();

    private enum InitializationState {
        NONE,
        IN_PROGRESS,
        COMPLETE
    }

    private static InitializationState initializationState = InitializationState.NONE;

    public synchronized static void initialize(@NonNull Context context) {
        if (initializationState == InitializationState.NONE) {
            initializationState = InitializationState.IN_PROGRESS;
            // addObserver must execute on the mainThread
            Handler mainHandler = new Handler(context.getMainLooper());
            mainHandler.post(new Runnable() {
                @Override
                public void run() {
                    try {
                        ProcessLifecycleOwner.get().getLifecycle().addObserver(new ProcessObserver());
                        initializationState = InitializationState.COMPLETE;
                    } catch (NoClassDefFoundError e) {
                        initializationState = InitializationState.NONE;
                        Logger.e(TAG,"Class 'ProcessLifecycleOwner' not found. The tracker can't track lifecycle events.");
                    }
                }
            });
        }
    }

    private ProcessObserver() {}

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onEnterForeground() {
        Logger.d(TAG, "App enter foreground");
        try {
            Map<String, Object> notificationData = new HashMap<String, Object>();
            notificationData.put("isForeground", Boolean.TRUE);
            NotificationCenter.postNotification("SnowplowLifecycleTracking", notificationData);
        } catch (Exception e) {
            Logger.e(TAG, "Method onEnterForeground raised an exception: %s", e);
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    public void onEnterBackground() {
        Logger.d(TAG, "App enter background");
        try {
            Map<String, Object> notificationData = new HashMap<String, Object>();
            notificationData.put("isForeground", Boolean.FALSE);
            NotificationCenter.postNotification("SnowplowLifecycleTracking", notificationData);
        } catch (Exception e) {
            Logger.e(TAG, "Method onEnterBackground raised an exception: %s", e);
        }
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/session/ProcessObserver.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.session

import android.content.Context
import android.os.Handler
import androidx.annotation.RestrictTo
import androidx.lifecycle.*
import com.snowplowanalytics.core.tracker.Logger
import com.snowplowanalytics.core.utils.NotificationCenter

import kotlin.Any
import kotlin.Exception
import kotlin.String


@RestrictTo(RestrictTo.Scope.LIBRARY)
class ProcessObserver private constructor() : DefaultLifecycleObserver {
    private enum class InitializationState {
        NONE, IN_PROGRESS, COMPLETE
    }

    override fun onStart(owner: LifecycleOwner) {
        super.onStart(owner)
        Logger.d(TAG, "App enter foreground")
        try {
            val notificationData: MutableMap<String, Any> = HashMap()
            notificationData["isForeground"] = true
            NotificationCenter.postNotification("SnowplowLifecycleTracking", notificationData)
        } catch (e: Exception) {
            Logger.e(TAG, "Method onEnterForeground raised an exception: %s", e)
        }
    }

    override fun onStop(owner: LifecycleOwner) {
        super.onStop(owner)
        Logger.d(TAG, "App enter background")
        try {
            val notificationData: MutableMap<String, Any> = HashMap()
            notificationData["isForeground"] = false
            NotificationCenter.postNotification("SnowplowLifecycleTracking", notificationData)
        } catch (e: Exception) {
            Logger.e(TAG, "Method onEnterBackground raised an exception: %s", e)
        }
    }

    companion object {
        private val TAG = ProcessObserver::class.java.simpleName
        private var initializationState = InitializationState.NONE
        
        @JvmStatic
        @Synchronized
        fun initialize(context: Context) {
            if (initializationState == InitializationState.NONE) {
                initializationState = InitializationState.IN_PROGRESS
                // addObserver must execute on the mainThread
                val mainHandler = Handler(context.mainLooper)
                mainHandler.post {
                    try {
                        ProcessLifecycleOwner.get().lifecycle.addObserver(ProcessObserver())
                        initializationState = InitializationState.COMPLETE
                    } catch (e: NoClassDefFoundError) {
                        initializationState = InitializationState.NONE
                        Logger.e(
                            TAG,
                            "Class 'ProcessLifecycleOwner' not found. The tracker can't track lifecycle events."
                        )
                    }
                }
            }
        }
    }
}
